### ISSUE TITLE
nmf parallel hangs forever

### ISSUE THREAD

**Issue by rronen** [renozao/NMF#57]
Hello,

```
nmf(rmatrix(10,10), 5, nrun=2)
```

This hangs forever on my unix machine whereas it takes <2 sec on my OS X. I have downgraded package versions of the relevant packages on the unix to match the mac:
```
 [1] doParallel_1.0.8    iterators_1.0.7     foreach_1.4.2      
 [4] NMF_0.20.5          synchronicity_1.1.4 bigmemory_4.4.6 
```
so I'd guess it's probably not a version issue. One difference is on the mac I have R 3.0.2 and on unix it's 3.1.2; so far I've avoided a complete re-install but will consider if you think it's necessary.

Here's the complete `SessionInfo()`:
```
R version 3.1.2 (2014-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] doParallel_1.0.8    iterators_1.0.7     foreach_1.4.2      
 [4] NMF_0.20.5          synchronicity_1.1.4 bigmemory_4.4.6    
 [7] BH_1.58.0-1         bigmemory.sri_0.1.3 Biobase_2.26.0     
[10] BiocGenerics_0.12.1 cluster_2.0.3       rngtools_1.2.4     
[13] pkgmaker_0.22       registry_0.3       

loaded via a namespace (and not attached):
 [1] codetools_0.2-14   colorspace_1.2-6   compiler_3.1.2     digest_0.6.8      
 [5] ggplot2_1.0.1      grid_3.1.2         gridBase_0.4-7     gtable_0.1.2      
 [9] magrittr_1.5       MASS_7.3-44        munsell_0.4.2      plyr_1.8.3        
[13] proto_0.3-10       RColorBrewer_1.1-2 Rcpp_0.12.1        reshape2_1.4.1    
[17] scales_0.3.0       stringi_1.0-1      stringr_1.0.0      tools_3.1.2       
[21] xtable_1.7-4      
```

**Comment by rronen** [comment]
update: also tried being more explicit + setting shared memory to FALSE (how it is set on the OS X where things work); but, no change - still hanging.

```
nmf(rmatrix(10,10), 5, nrun=2, .opt=list(debug=F, parallel=T, shared.memory=F, pbackend='par'))
```

**Comment by renozao** [comment]
Can't reproduce the error on my Ubuntu machine. Have you tried using the latest version of NMF on CRAN (0.20.6)?

**Comment by rronen** [comment]
Yes, this is what I started with (then downgraded back to 0.20.5 since that works okay on my mac). Everything works fine if I set `parallel =FALSE` but it makes the necessary number of iterations prohibitively long.

**Comment by rronen** [comment]
Thanks @renozao for trying to reproduce, is there any additional info I can dig up that might be useful in debugging or diagnosing this? Your help is much appreciated.

**Comment by renozao** [comment]
Want to exclude issues due to dependencies.
Can you please try re-install the bigmemory-related packages:

```r
install.pacakges(c('synchronicity', 'bigmemory.sri', 'bigmemory'))
```

Test, and if not working then try re-installing `digest`:

```r
install.packages('digest')
```

It solved a problem for me once. Not sure why, possibly due to system updates.

**Comment by rronen** [comment]
Hi @renozao ,

I've re-installed as you suggested (including `digest`), and no luck still this command hangs forever.
```
nmf(rmatrix(10,10), 5, nrun=2, .opt=list(debug=F, parallel=T, shared.memory=F, pbackend='par'))
```

`sessionInfo()` now is:
```
> sessionInfo()
R version 3.1.2 (2014-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] doParallel_1.0.8    iterators_1.0.7     foreach_1.4.2      
 [4] NMF_0.20.5          synchronicity_1.1.4 bigmemory_4.5.8    
 [7] bigmemory.sri_0.1.3 Biobase_2.26.0      BiocGenerics_0.12.1
[10] cluster_2.0.3       rngtools_1.2.4      pkgmaker_0.22      
[13] registry_0.3       

loaded via a namespace (and not attached):
 [1] codetools_0.2-14   colorspace_1.2-6   compiler_3.1.2     digest_0.6.8      
 [5] ggplot2_1.0.1      grid_3.1.2         gridBase_0.4-7     gtable_0.1.2      
 [9] magrittr_1.5       MASS_7.3-44        munsell_0.4.2      plyr_1.8.3        
[13] proto_0.3-10       RColorBrewer_1.1-2 Rcpp_0.12.1        reshape2_1.4.1    
[17] scales_0.3.0       stringi_1.0-1      stringr_1.0.0      tools_3.1.2       
[21] xtable_1.7-4 
```

Also I've noticed that running the same command with rank=2 (rather than 5) setting `parallel=T` does work ok (both return very quickly with `parallel=F`).

**Comment by renozao** [comment]
This is what I did, on R 3.2.2, and it works fine. 
Beside the base R packages (tools, compiler, etc...) and the Biobase packages, there not much differences with your settings.
Are you able to try with R-3.2.2? Re-installing Rcpp maybe (really trying all mild-irrational stuff here)?

```r
library(NMF)
set.seed(12345)
nmf(rmatrix(10,10), 5, nrun=2, seed = 678, .opt=list(debug=F, parallel=T, shared.memory=F, pbackend='par'))
```

```r
> sessionInfo()
R version 3.2.2 (2015-08-14)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.3 LTS

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_ZA.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_ZA.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_ZA.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_ZA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] doParallel_1.0.8    iterators_1.0.7     foreach_1.4.2      
 [4] NMF_0.20.6          synchronicity_1.1.4 bigmemory_4.5.8    
 [7] bigmemory.sri_0.1.3 Biobase_2.28.0      BiocGenerics_0.14.0
[10] cluster_2.0.3       rngtools_1.2.4      pkgmaker_0.22      
[13] registry_0.3       

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.1        magrittr_1.5       MASS_7.3-44        munsell_0.4.2     
 [5] colorspace_1.2-6   xtable_1.7-4       gridBase_0.4-7     stringr_1.0.0     
 [9] plyr_1.8.3         tcltk_3.2.2        tools_3.2.2        grid_3.2.2        
[13] gtable_0.1.2       digest_0.6.8       RColorBrewer_1.1-2 reshape2_1.4.1    
[17] ggplot2_1.0.1      codetools_0.2-14   stringi_1.0-1      compiler_3.2.2    
[21] scales_0.3.0       proto_0.3-10
```

**Comment by kaneplusplus** [comment]
@renozao do you create named, shared mutexes when nmf is performed in parallel? If so, and R crashes it is possible that they could not be cleaned up correctly. @rronen you'll be able to see them on linux in the /dev/shm directory.

**Comment by rronen** [comment]
Thanks @kaneplusplus,  rather than crash what I see is R hanging (what seems to be) indefinitely after I issue the `nmf()` command with `.opt=list(parallel=T)`. Interestingly, this seems to be somehow related to the rank; using rank of 2 it returns but 5 causes hanging.

**Comment by kaneplusplus** [comment]
Can you run the following in a terminal window so we can see the shared resources?
```
ls -lah /dev/shm
```
If you see resources that end with _counter then I have a problem.

**Comment by rronen** [comment]
Indeed, `ls -lah /dev/shm/` lists 171 files, most of which end with `_counter` or `_counter_mutex`. 

I wonder if some/many of these exist because I killed (ctrl+c) many tries of NMF that were hanging (while testing things on this thread).

**Comment by kaneplusplus** [comment]
Shoot, I can verify this. This is my mistake. Now, let's see if we can get the example to work, with the memory leak. Please exit R, remove all of those shared resources, and see if the example works.

**Comment by kaneplusplus** [comment]
OK, I think I found the problem and it's mine... Give me an hour or so and I'll report back.

**Comment by rronen** [comment]
Okay, deleted everything that ends with `_counter` or `_counter_mutex` from `/dev/shm/`; 

Re-started, ran:
```
library(NMF)
set.seed(12345)
nmf(rmatrix(10,10), 5, nrun=2, seed=678, .opt=list(debug=F, parallel=T, shared.memory=F, pbackend='par'))
```

I have two threads working at 100% for a few minutes, seems there is no change. Is this what you meant? Killing it now. As far as I can see there were no files ending with `_counter` or `_counter_mutex` in `/dev/shm`; also none now after I killed the `nmf()` command.

**Comment by kaneplusplus** [comment]
@rronen you probably need to remove all of the associated shared resources. The ```big.matrix``` objects are not being cleaned up but I'm having trouble figuring out why this might be happening.

@renozao are you creating big.matrices in C? I've just checked in some updates but I want to make sure there couldn't be a problem with a memory leak.

**Comment by kaneplusplus** [comment]
OK, I think I found the problem. The example uses "doMC" which forks for parallel processes. When I run
```{r}
library(bigmemory)
library(doMC)
registerDoMC(cores=2)
a = big.matrix(nrow=3, ncol=3, shared=TRUE)
desc = describe(a)
foreach (i=1:3) %dopar% {
  m = attach.big.matrix(desc)
  m[i,1] = i
  NULL
}
```
I get a resource leak because the destructor is not called (depite the fact that I've specified TRUE for the onexit parameter). If I add the line ```rm(m); gc()``` before ```NULL``` I don't see the problem. My guess is that forked process cleans itself up a little bit different than the parent.

It looks to me like there are 2 obvious solutions but I'm happy to hear others.
1. The user of synchronicity and bigmemory have to make sure they clean up shared resources with ```rm``` and ```gc```
2. The process that creates a shared matrix is responsible for destroying it.

The second option looks better to me. @renozao if this also seems reasonable I'll check around to make sure there aren't other people that have objections.

Thanks for bringing this to my attention.

**Comment by renozao** [comment]
@rronen if you use option `shared.memory=FALSE`, then `nmf` should not use `bigmemory` or `synchronicity` at all.
You could check this by uninstalling `bigmemory` and re-run your sample code.

@kaneplusplus I create the `big.matrix` object in R using the code:

```r
bigmemory::as.big.matrix(init, type='double', shared=TRUE)
```

with `init` a single numeric `NA` or a numeric matrix full of zeros.

The mutexes are created using the code:

```r
synchronicity::describe(synchronicity::boost.mutex())
```

But again, in the example @rronen posted, neither `big.matrix` nor `boost.mutex` should be called.

@rronen, you said that using `parallel = FALSE` works. Can you try the following, which uses another parallel backend:

```r
library(NMF)
# create data
set.seed(12345)
x <- rmatrix(10,10)

# run not in parallel (should work)
# NOTE: .pbackend is a separate parameter, not in the .opt list
nmf(x, 5, nrun=2, seed=678, .opt=list(debug=F, parallel=FALSE, shared.memory=F)
     , .pbackend='par')

# run using SNOW backend
library(doSNOW)
nmf(x, 5, nrun=2, seed=678, .opt=list(debug=F, parallel=TRUE, shared.memory=F)
     , .pbackend='psock')

# same thing using shared memory
nmf(x, 5, nrun=2, seed=678, .opt=list(debug=F, parallel=TRUE, shared.memory=TRUE)
     , .pbackend='psock')
```

**Comment by kaneplusplus** [comment]
So, just to summarize it looks like there are 3 separate issues.
1. nmf hangs - @renozao and I have not been able to reproduce this using parallel and/or shared memory.
2. shared big.matrix objects have a resource leak - my example of this is shown above.
3. synchronicity has a resource leak - the example is similar to the bigmemory example above.

Apologies for the shared resource digression - I was wondering if it was hanging trying to to allocate them. I will file bug reports for 2 and 3 in bigmemory and synchronicity respectively and handle it separately.

**Comment by d3d1y** [comment]
I hope I'm just doing this wrong. But I'm wondering if the problem I'm having might be related to this issue of nmf parallel hanging forever. I've used NMF happily for years, but this is the first time that I've tried using it with bigmemory. I have a big'ish matrix to factor (303x9550). It wasn't running using my old settings, so I tried sharedmemory. Now it runs up to 6% and stalls.

Here is my session info:
> sessionInfo()
R version 3.1.1 (2014-07-10)
Platform: x86_64-unknown-linux-gnu (64-bit)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] NMF_0.20.5          synchronicity_1.1.0 bigmemory_4.5.8    
 [4] bigmemory.sri_0.1.3 Biobase_2.26.0      BiocGenerics_0.12.0
 [7] cluster_1.15.3      rngtools_1.2.4      pkgmaker_0.22      
[10] registry_0.2       

loaded via a namespace (and not attached):
 [1] codetools_0.2-9    colorspace_1.2-4   digest_0.6.4       doParallel_1.0.8  
 [5] foreach_1.4.2      ggplot2_1.0.0      grid_3.1.1         gridBase_0.4-7    
 [9] gtable_0.1.2       iterators_1.0.7    MASS_7.3-35        munsell_0.4.2     
[13] plyr_1.8.1         proto_0.3-10       RColorBrewer_1.0-5 Rcpp_0.11.3       
[17] reshape2_1.4       scales_0.2.4       stringr_0.6.2      tools_3.1.1       
[21] xtable_1.7-4  

Here is my nmf call:
> dim(V.matrix)
[1]  303 9550
> res <- nmf(V.matrix, 2, nrun=500, .opt=list(debug=F, verbose=T, track=T, parallel=T, shared.memory=T, pbackend=12))
NMF algorithm: 'brunet'
Multiple runs: 500
Mode: parallel (31/32 core(s))
Runs: |===                                               |   6%
<...stalls...>

**Comment by andrewjkwok** [comment]
I was wondering if these issues were ever fully resolved as I am trying to the package out and running into a similar problem where the nmf function hangs. Specifically, it happens when I try to run `nmf` for more than one rank, when I run with `.option='vd'` I see that the run hangs on the second rank being tested, and once it hangs, I have to restart my Rstudio session until I can get it working again.

**Comment by Yunuuuu** [comment]
The same problem, always hand in the second rank
